### ISSUE TITLE
Snippet editor progress bars accessibility

### ISSUE THREAD

**Issue by afercia** [Yoast/YoastSEO.js#466]
Moving here issue wrongly opened on wordpress-seo, see https://github.com/Yoast/wordpress-seo/issues/4135

As also @diedexx commented, see https://github.com/Yoast/YoastSEO.js/pull/448#issuecomment-213405785 the progress bars value is not announced by VoiceOver. Reporting here the screenshot from #448 :

![236550f0-0896-11e6-9852-d9c5c8a5081f](https://cloud.githubusercontent.com/assets/1682452/14865764/89934cf0-0cc0-11e6-9904-88699b031ec5.png)

While the implementation is technically correct, seems not all browsers / screen readers combinations get it right. The native HTML5 `<progress>` element with `value` and `max` attributes should suffice because it already carries the necessary semantics and information. But, while webkit supports the `<progress>` element, probably VoiceOver doesn't yet. In my testing, to make VoiceOver work with both Safari and Chrome it is necessary to add a `role="progressbar"` and the aria attributes @diedexx mentioned.

The reason why it announces "indeterminate" is because it needs the "max" value set as attribute (not as property) to be present since the beginning.

Worth noting the value announced is the result of the internal calculation screen readers do based on the min, max and current value so we can't solve the rounding issue and values will be announced with decimals, unless we base all on a scale from 0 to 100, maybe. Needs more testing with other screen readers.

Screenshot of my testing so far:

<img width="1440" alt="screen shot 2016-04-27 at 19 09 50" src="https://cloud.githubusercontent.com/assets/1682452/14865835/f2537a8a-0cc0-11e6-8a97-3fc57fc53d92.png">

Still to evaluate:
- not sure if setting values as properties is OK, for example `element.value = value` same for "max"
- add an aria label to explain "what" these progress bar are about
- maybe add an aria-live region to announce the progress value when users stop typing

**Comment by afercia** [comment]
Note:
https://www.w3.org/TR/wai-aria/states_and_properties#aria-valuenow
> For progressbar elements and scrollbar elements, assistive technologies SHOULD render the value to users as a percent, calculated as a position on the range from aria-valuemin to aria-valuemax if both are defined

https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_progressbar_role
> Note: Assistive technologies generally will render the value of aria-valuenow as a percent of the range between the value of aria-valuemin and aria-valuemax, unless aria-valuetext is specified. It is best to set the values for aria-valuemin, aria-valuemax, and aria-valuenow in a manner that is appropriate for this calculation.

This needs more investigation, as far as I see the implementation across screen readers varies, for example some read out "100%" and others "65". Going to build a simplified test case on Codepen.

Possible alternative to consider (and test): a human readable text alternative (e.g. Bad, OK, Good)
https://www.w3.org/TR/wai-aria/states_and_properties#aria-valuetext

**Comment by afercia** [comment]
I'm getting all sort of results trying to make progress bars read out correctly across browsers and screen readers. The current support seems very inconsistent. In some cases even CSS properties affect the result. Specifically, Safari doesn't like `-webkit-appearance: none;` and that's the reason why in our Snippet Editor it reads out "indeterminate".
I'd greatly appreciate some testing to confirm what I'm experiencing here, please see this simplified test case on Codepen:
 http://s.codepen.io/afercia/debug/xVyXPo

**Comment by atimmer** [comment]
@afercia Would `<meter>` help us out here or has it just the same problems. I think for semantic reasons `<progress>` was inaccurate when I first implemented this. But they are probably implemented the same in webkit.

**Comment by afercia** [comment]
@atimmer yep, `<progress>` is more for a task completion and `<meter>` maybe would be semantically more appropriate. It depends on what kind of information we want to convey. It is based on the number of characters (65 and 156) but it also changes color from orange to green. At the same time the bar suggests an ongoing process from 0% to 100%. So maybe we should first decide what we want to communicate from an SEO perspective:
- the number of characters left (I guess no?)
- a 0/100 percentage (filling the fields is a "task" in a certain sense)
- just "too short" and "OK"? this could be done with `aria-valuetext`
Also, I'm not sure about the `<meter>` element browsers support. The `<progress>` support is already problematic and inconsistent across browsers so I guess `<meter>` support is even worse but this would require some testing. I'd say it needs a decision about the message we want to convey as first thing :)

**Comment by atimmer** [comment]
Yes, `<meter>` has worse browser support: http://caniuse.com/#feat=meter
Compare to: http://caniuse.com/#feat=progress

So we can ignore my comment for now.

Another question is whether or not the progress bars should even be read out by screenreaders. The same information is also conveyed by the suggestions generated by the assessments. The progress bar is only a visual guide to help get to a better title/description.

@omarreiss What do you think?

**Comment by omarreiss** [comment]
@atimmer I agree. People using screen readers probably need a different information density. I guess triggering aria events on changes in the analysis results would provide the user with the needed amount of feedback. This was one of the things discussed with @afercia at WC Torino, but I am not sure if there's an issue for that yet.

**Comment by afercia** [comment]
There is, but on the wpseo repo :) see https://github.com/Yoast/wordpress-seo/issues/4288

**Comment by afercia** [comment]
So I'd propose for now to hide the progress bars from assistive technologies using `aria-hidden` (hopefully it will work on `<progress>` elements) and move them outside of the `<label>` elements.

Labels should only contain one labelable element. That also because screen reader may read out the progress bars when focusing the input field, as if they were real labels.

![progress bars inside labels nvda](https://cloud.githubusercontent.com/assets/1682452/15386757/d5f1796a-1da9-11e6-8938-74ccfcabc207.png)
